### ISSUE TITLE
TextEdit: wrap is not implementing

### ISSUE THREAD

**Issue by neikeq** [godotengine/godot#3985]
**Issue description** (what happened, and what was expected):

_TextEdit_ has method `set_wrap(bool)` which is a setter for the variable [wrap](https://github.com/godotengine/godot/blob/master/scene/gui/text_edit.h#L207). This variable is never used, so I assume this option was never implemented (or was removed).

This should be either implemented or the method should be removed to avoid confusion (like [this one](http://godotengine.org/qa/838/textedit-set_wrap-doesnt-seem-to-do-anything-thoughts)).

**Comment by reduz** [comment]
It used to work.. wonder what happened..

**Comment by neikeq** [comment]
I went through all [scene/gui/text_edit.cpp](https://github.com/godotengine/godot/commits/master/scene/gui/text_edit.cpp) commits, and none seem to have touched this option. Maybe it was removed before going open source?

**Comment by akien-mga** [comment]
The ``wrap`` argument does not seem to be used anywhere in ``text_edit.cpp``.

**Comment by acalc** [comment]
Reporting in that this is still an issue in the 2.1 beta release.

**Comment by YeOldeDM** [comment]
Bump! set_wrap still not working in 2.1.stable.official

**Comment by Qualitymix** [comment]
Bump.  This is a real bummer!

**Comment by reduz** [comment]
PR for this welcome

**Comment by akien-mga** [comment]
Note that it should ideally also be added in the script editor (and maybe made configurable, so that you can e.g. wrap at 80 chars) after it has been implemented in TextEdit.

**Comment by Qualitymix** [comment]
I noticed this is not currently in 2.1.1. Will it make its way there?



Sent from TypeApp<http://www.typeapp.com/r>

On Oct 9, 2016, at 3:00 PM, "Rémi Verschelde" <notifications@github.com<mailto:notifications@github.com>> wrote:

Note that it should ideally also be added in the script editor (and maybe made configurable, so that you can e.g. wrap at 80 chars) after it has been implemented in TextEdit.

**Comment by akien-mga** [comment]
> I noticed this is not currently in 2.1.1. Will it make its way there?

It would have to be implemented first, and 2.1.1 is not released.

**Comment by Qualitymix** [comment]
Got it.  I downloaded and compiled from master branch.  Are there any roadblocks to implementing this ? Or is it just a matter of time and priorities?


Sent from TypeApp<http://www.typeapp.com/r>

On Oct 10, 2016, at 12:00 PM, "Rémi Verschelde" <notifications@github.com<mailto:notifications@github.com>> wrote:

I noticed this is not currently in 2.1.1. Will it make its way there?

It would have to be implemented first, and 2.1.1 is not released.

**Comment by reduz** [comment]
The only roadblock is that no one offered to do it, same as code folding.
So if anyone wants to contribute this work be welcome

**Comment by Qualitymix** [comment]
If someone could point to where in the source code the textedit node is located I could take a look at it.  C++ right? I might be able to help.



Sent from TypeApp<http://www.typeapp.com/r>

On Oct 10, 2016, at 12:10 PM, Juan Linietsky <notifications@github.com<mailto:notifications@github.com>> wrote:
The only roadblock is that no one offered to do it, same as code folding.
So if anyone wants to contribute this work be welcome

**Comment by reduz** [comment]
great! :D

source code of text edit is scene/gui/text_edit.h,.cpp

everything is in there

**Comment by Qualitymix** [comment]
Drawing blanks with this.   The C++ code is hard for me to grasp, not having written it myself.   I suck at reading others' code. :(

**Comment by Fiona** [comment]
Oh no! I just spent all day investigating the possibility of writing a simple text editor for my tablet today using Godot and just came across this issue. Godot was absolutely perfect for the job but then I discovered set_wrap didn't work and my plans were completely ruined.

I've had a glance at text_edit.cpp but it's largely uncommented black magic it seems.

I would be willing to add a $50 bounty to this, though I know that's massively low-balling the work required.

**Comment by DenisComtesse** [comment]
It would be great if this could be fixed. TextEdit is a very useful node, but it's quite annoying that wrapping doesn't work.

**Comment by penguinroad** [comment]
Yep. Definitelly wait for this one too. Please work on this

**Comment by YeOldeDM** [comment]
yes, please fix this.  With text wrapping, an in-game wiki system can be made complete :sunglasses:

**Comment by willnationsdev** [comment]
Ok, I'm going to go ahead and try to tackle this. Specifically because I have about 3 different tool concepts I wanna work on for Godot and ALL of them would require textedit wrapping to have the right feel.

Symptoms I've noticed (in summary):

- set_wrap is not effective.
- set_max_chars is not effective.
- If you hold down shift and highlight the content of the TextEdit, the section will be highlighted, but then the remaining contents of the TextEdit will begin flashing grey/white so long as you aren't holding down shift anymore.

I'll try to work on these. I also plan to implement a max_line_chars variable so that you can specify a max length per line.

**Comment by reduz** [comment]
Feel free to contribute a fix!

**Comment by willnationsdev** [comment]
@reduz Is there a reason you know of why text_edit.cpp::_update_line_cache runs through the lines checking for color regions two completely separate times (lines 151-208)? It seems like it would just end up adding two separate color region info objects for each actual color region...

**Comment by reduz** [comment]
Color regions can span multiple lines for comments, line constants, etc.

**Comment by pwab** [comment]
TextEdit should also have an auto_height when wrap is enabled. See #9746.

**Comment by willnationsdev** [comment]
TextEdit is storing its underlying data as a mutable Vector<String> (one String for each line). Does anyone have any suggestions on how best to tackle the cascading repositioning of strings as you resize the window / insert text? My first thought would just be to move things to lines below/above as the window contracts/expands; however, doing that would essentially be pretending that `\n` exists where it doesn't (it would reflect in an increased/decreased count for line numbers which I would want to avoid).

Place of interest is /scene/gui/text_edit.cpp, line 2776, method _base_insert_text

It seems like the most appropriate way to handle it is to write an algorithm that...

1. Connects preinsert_text to the next text on the insertion line.
2. Checks the current line and detects the starting position of the first word to go beyond the visible width.
3. Adds a new line to the editor WITHOUT adding a new line (best way to handle that?) and write the remaining text on the current line.
4. Repeats steps 1 and 2 for each new line created until no more text goes beyond the visible width.
5. Reconnects the postinsert_text.
6. Repeats step 4.

**Comment by reduz** [comment]
Probably the best is to use, for each line, a sub-line array (starting
positions of sub-lines, like, Vector<int> that is cached and re-computed on
screen resize (width) or line modification.

**Comment by willnationsdev** [comment]
Trying to think of how that would work in practice, the actual drawing of it in `NOTIFICATION_DRAW`. I figure I'd have to amend things so that the TextEdit keeps track of how many ACTUAL lines it has versus the number of VISUAL lines (actual + sub-lines), and all of the drawing operations would have to work off of the visual lines while things like highlighting a breakpoint/marked row shows every line and its sub-lines. That way you can reposition the `font->draw_char` parameters to target the visual position and whatnot. Geez, this is gonna be more complicated than I thought. XD

**Comment by willnationsdev** [comment]
Update: I have implemented the following bug fixes / features:

1. block_caret altering subsequent font color change (bug fix)
2. max_chars not working properly (bug fix)
3. Adding a max_line_chars (and associated getter/setter methods)
4. Adding signals for both max_chars and max_line_chars (max_chars_reached and max_line_chars_reached, respectively) notifying the user when an attempt is made to insert beyond the limitation and which line.

I am now moving on to the actual wrap implementation (already have an initial implementation caching the appropriate string indexes during insertion for later wrapping).

**Comment by willnationsdev** [comment]
Another update:

I am realizing the large task that involves word wrapping. XD It took a while to really figure out where everything is happening / what everything does, but I think I'm getting the hang of it.

So far, I have text showing up in a wrapped fashion as it moves off the screen. I'm going to have to start messing with the more low-level calculations however, so that they can return "wrapped" coordinates or use "wrapped" coordinates when wrap is toggled (scrollbars, viewport adjustments, selection box highlighting). I also still need to teach the cursor to adjust its writing position based on the presence of sub-rows when wrap is toggled since it currently will have new lines writing on top of the previous lines' wrapped rows of text. This'll take me a while probably as there are a lot of changes to make.

**Comment by stubbsy345** [comment]
Thanks so much for doing this my game is desperate for word wrapping to work with text edit. It's a problem I have been having for a while. How are you getting on with the task?

**Comment by willnationsdev** [comment]
Well, For the last week I've been on vacation with family (programming forbidden) and before that i was rushing to get work stuff done, so it's pretty much in the state mentioned above. You can review the progress at willnationsdev/godotengine on the text-edit-wrap-fix branch.

**Comment by willnationsdev** [comment]
Another quick update: I've managed to get the text to display in the proper position on subsequent lines (no longer writes ON TOP of previous wrapped lines, but moves down to accommodate them). I've also disabled the horizontal scrollbar when wrap is enabled. Currently working on getting the viewport adjustment and vertical scrollbar to respond appropriately to wrapped lines so that they don't start getting drawn off the edge of the TextEdit.

**Comment by stubbsy345** [comment]
Hi willnationdev, just a quick question, will your textwrap fix also be applicable to godot 3.0?

**Comment by willnationsdev** [comment]
@stubbsy345 Hey, sorry, I saw this in an email notification, and then forgot to respond.

Unfortunately, I don't think I'll be able to get it in for the 3.0 release. I MIGHT be able to isolate the max_chars fixes / features I mentioned if I can isolate them out of my branch, but I've been having really weird problems in getting my version of godot to build correctly as of late.

I was working on a different project for the last month - something I could do without having to compile the engine from source, cause I was starting to get burnt out / frustrated. I wanna try re-tackling my engine compilation issues though, so hopefully I can resume work on this.

**Comment by stubbsy345** [comment]
Missed this as well. Awesome. If I can aid in any way let me know.

**Comment by willnationsdev** [comment]
Yeah, definitely. I wanna finish up the project I've been working on, and then once that's done, I'll get a repo going again with the edit `text_edit.cpp/.h` files. I'll update you again once it's up and we can work together on it. Sound good?

**Comment by stubbsy345** [comment]
I'm not sure my C++ is up to a standard of being of a great deal of use but if there is work I can do in another regard to help I am happy to. I was wondering, from the work you have already done, is it possible to extract just a very simple text wrapping text edit node. Many people who want the user to be able input text that wraps don't need all of the functions that text edit provides - only for the text to wrap and scroll to adapt as the length of the text increases, or possibly not even this, the node could be placed inside of a scroll container, similar to the way Vbox's can be placed inside such. Is this possible? Or is it just as easy to implement it for the current text edit node as it is now.

**Comment by willnationsdev** [comment]
@stubbsy345 Ah, sorry again for not getting back to you. The scrolling functionality that remains to be written is still necessary (can't be funneled into a ScrollContainer) because if I don't do it, the node literally starts drawing code OFF the bottom of the node into the space below it. I had to close the repository because I screwed up the commit stuff, but I've saved a copy of my progress on the .h/.cpp files and emailed them to myself, so I'll be recreating the repo some time down the line. Will post an update here once I resume work on the project, but it'll be a while. Spent most of the last month working in simpler engine modifications for the recent beta release, so now I'm getting back on my plugins.

**Comment by blurymind** [comment]
Hope this gets fixed, because I need it.
Still doesn't work in Godot 3

**Comment by Supergeek** [comment]
When I was playing around with Godot, this is the missing feature that made me reluctantly walk away. This is one of those features that, when you need it, you really can't do without.

**Comment by willnationsdev** [comment]
@blurymind @Supergeek there is already a PR for this ready to go. It's scheduled for the 3.1 release.

**Comment by lectronice** [comment]
Hey, sorry if I'm posting this in the wrong place, I'm not sure if I should open a new issue or not.

Text wrap now works in 3.1 alpha 1, but accented characters (é, è, à, etc.) in wrapped lines randomly disappear. Expanding the window to display single lines make them reappear. I've tried various font and TextEdit settings but this odd behavior only seem to happen when text warp is enabled.

**Comment by akien-mga** [comment]
> Hey, sorry if I'm posting this in the wrong place, I'm not sure if I should open a new issue or not.

Please open a new issue yes.

**Comment by Lafender** [comment]
hey so I can someone direct me on how to get godot 3.1 alpha? i'm using a ubuntu.

also i tried implementing my own version of autowrap in version 3.06 with the following code, and it also didn't work. is that related to why it wasn't implemented possibly?

![auto-wrap-godot02](https://user-images.githubusercontent.com/11316682/50049378-a1c91780-00b2-11e9-9812-afa7acae5226.png)


> Hey, sorry if I'm posting this in the wrong place, I'm not sure if I should open a new issue or not.
> 
> Text wrap now works in 3.1 alpha 1, but accented characters (é, è, à, etc.) in wrapped lines randomly disappear. Expanding the window to display single lines make them reappear. I've tried various font and TextEdit settings but this odd behavior only seem to happen when text warp is enabled.

**Comment by lectronice** [comment]
Watch the official blog for alpha releases. Alpha 3 is out: https://godotengine.org/article/dev-snapshot-godot-3-1-alpha-3

**Comment by lauren-emily75** [comment]
Ran into this issue and so downloaded 3.1 alpha after reading this. Wrapping now works but when text moves to the next line I start getting hundreds of errors like this.
![image](https://user-images.githubusercontent.com/23086413/54126902-82672f80-4400-11e9-8f42-eba029c5351f.png)

Any idea what could be causing this? My game still seems to function despite the errors, just wondering if this is actually a problem

**Comment by pwab** [comment]
I also got those error lines a while back. But I wasn't able to reproduce this now with a minimal project on **Godot 3.1 RC 2**.

Here is my test project: [wrap_test.zip](https://github.com/godotengine/godot/files/2952478/wrap_test.zip)

This is how it looks:
![wrap_test](https://user-images.githubusercontent.com/19271185/54129136-e80de880-440d-11e9-8a73-1ba51d8629e7.gif)


@lauren-emily75: Could you provide a minimum project where this happens?

**Comment by lauren-emily75** [comment]
Hi @pwab , I've just made a simple project with a control node and text edit node as a child. I haven't attached any scripts, just selected wrap enabled in the inspector. When entering text I start getting errors when the text moves to a new line and then from there every time I type a character.

I downloaded and tried yours, entering the text myself and like you said, there are no errors so not really sure what's wrong with mine. I have tried reproducing yours myself and am still getting the same errors

**Comment by lauren-emily75** [comment]
@pwab turns out I was using an old alpha version. Downloaded the latest and there is no issue!

**Comment by pwab** [comment]
@lauren-emily75 glad to hear this! 😃

**Comment by byrro** [comment]
If you came here because `wrap_enabled` or `set_wrap_enabled` is still not working in 2022: make sure you set the `text` property **after** you set `wrap_enabled`, otherwise it will silently fail.

**Comment by Calinou** [comment]
> If you came here because `wrap_enabled` or `set_wrap_enabled` is still not working in 2022: make sure you set the `text` property **after** you set `wrap_enabled`, otherwise it will silently fail.

This sounds like a bug which should be reported in a separate issue (with a minimal reproduction project attached).

**Comment by byrro** [comment]
@Calinou sure, will do it later today. Since this thread comes up ranked high in search results, I wanted to document it here while this isn't fixed.

**Comment by lesleyrs** [comment]
I don't see an open issue so I will create one, it's still happening in Godot 4 beta 10. I use an empty text edit and it does not wrap with the boundary wrap mode during runtime unless I add default text to it. I think that's all you need to reproduce. It's not really important but I really like using the placeholder text area instead.
